### PR TITLE
Blacklist Sandy Bridge on mesa from using geometry shaders.

### PR DESCRIPTION
**Pull request by Sonicadvance1**
[Review on Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3518)

### CODE REVIEW

File: Source/Core/VideoCommon/DriverDetails.h [hunk 1]
```diff
@@ -182,10 +182,10 @@ namespace DriverDetails
 		BUG_BROKENALPHATEST,
 
 		// Bug: Broken lines in geometry shaders
-		// Affected Devices: Mesa r600/radeonsi
+		// Affected Devices: Mesa r600/radeonsi, Mesa Sandy Bridge
 		// Started Version: -1
 		// Ended Version: -1
-		// Mesa inroduced geometry shader support for radeon devices and failed to test it with us.
+		// Mesa inroduced geometry shader support for radeon and sandy bridge devices and failed to test it with us.
```
Comment (BhaaLseN) [inline Source/Core/VideoCommon/DriverDetails.h]: Mind fixing that typo while you're at it? "introduced" with a T

**Comment by delroth** [comment]
Why?

**Comment by Sonicadvance1** [comment]
For the same reason that radeon is blacklisted.

File: Source/Core/VideoBackends/OGL/Render.cpp [hunk 2]
```diff
@@ -210,11 +210,21 @@ static void InitDriverInfo()
...
 		{
 			if (svendor == "nouveau")
+			{
 				driver = DriverDetails::DRIVER_NOUVEAU;
+			}
 			else if (svendor == "Intel Open Source Technology Center")
+			{
 				driver = DriverDetails::DRIVER_I965;
+				if (srenderer.find("Sandybridge") != std::string::npos)
+					family = 2;
```
Comment (delroth) [inline Source/Core/VideoBackends/OGL/Render.cpp]: Constants for families?

**Comment by sheepdestroyer** [comment]
Is the bug reported upstream?

**Comment by Sonicadvance1** [comment]
The bug is reported somewhere to mesa for both the radeon and Intel drivers.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-:-K8Uwo_7DgJ1_jhj1_MV:-144260951)*

**Comment by BhaaLseN** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoCommon/DriverDetails.h, line 55 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-K8Uxfwx85U8S1nD0qWM:-K8Uxfwx85U8S1nD0qWN:1249681164) ([raw file](https://github.com/dolphin-emu/dolphin/blob/104286dd26ba3fd48ad4c05ebf560e13e861c4a0/Source/Core/VideoCommon/DriverDetails.h#L55)):*
`enum class`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by Sonicadvance1** [comment]
*[Source/Core/VideoCommon/DriverDetails.h, line 55 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-K8Uxfwx85U8S1nD0qWM:-K8UyFR4f7VpPP3EfKJn:1521777950) ([raw file](https://github.com/dolphin-emu/dolphin/blob/104286dd26ba3fd48ad4c05ebf560e13e861c4a0/Source/Core/VideoCommon/DriverDetails.h#L55)):*
Won't work, it's being cast to a u32. Which can't be easily changed over to use the class of DriverDetails::Family because the OS X and Windows paths rely on it to be a u32.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by BhaaLseN** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoCommon/DriverDetails.h, line 55 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-K8Uxfwx85U8S1nD0qWM:-K8Uz9wqga59jUWLb3vw:-944690147) ([raw file](https://github.com/dolphin-emu/dolphin/blob/104286dd26ba3fd48ad4c05ebf560e13e861c4a0/Source/Core/VideoCommon/DriverDetails.h#L55)):*
`enum class` also supports specifying a type if you want to: `enum class Family : std::uint32_t` (which requires `cstdint` to be included)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by Sonicadvance1** [comment]
*[Source/Core/VideoCommon/DriverDetails.h, line 55 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-K8Uxfwx85U8S1nD0qWM:-K8Uzt7ADIYaAbk9b6mW:1215480953) ([raw file](https://github.com/dolphin-emu/dolphin/blob/104286dd26ba3fd48ad4c05ebf560e13e861c4a0/Source/Core/VideoCommon/DriverDetails.h#L55)):*
That still doesn't fix the issue and it won't compile.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by delroth** [comment]
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoCommon/DriverDetails.h, line 55 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-K8Uxfwx85U8S1nD0qWM:-K8V-AdxU4PZw37xiKL9:-1786158803) ([raw file](https://github.com/dolphin-emu/dolphin/blob/104286dd26ba3fd48ad4c05ebf560e13e861c4a0/Source/Core/VideoCommon/DriverDetails.h#L55)):*
Suggestion: Add UNKNOWN = -1 to family, replace our uses of -1 in the bugs array by UNKNOWN, change the type to the array to Family. That way it can be enum class :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by Sonicadvance1** [comment]
*[Source/Core/VideoCommon/DriverDetails.h, line 55 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-K8Uxfwx85U8S1nD0qWM:-K8V-VCrYasNoFgFfTq6:-499478304) ([raw file](https://github.com/dolphin-emu/dolphin/blob/104286dd26ba3fd48ad4c05ebf560e13e861c4a0/Source/Core/VideoCommon/DriverDetails.h#L55)):*
That won't fix the OS X and Windows path requiring it as a u32 and I don't know how to query the family on those OSes to make sure I choose the proper one so the DriverDetail bug is still active.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by aserna3** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoCommon/DriverDetails.h, line 55 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-K8Uxfwx85U8S1nD0qWM:-K8V1U-7MZapH2MFP5H_:1861298409) ([raw file](https://github.com/dolphin-emu/dolphin/blob/104286dd26ba3fd48ad4c05ebf560e13e861c4a0/Source/Core/VideoCommon/DriverDetails.h#L55)):*
I think @lioncash has an OS X system. Maybe he would know the solution here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by sheepdestroyer** [comment]
Should specific driver bugs reports be referenced with these workarounds, so that progress be tracked and the blacklist lifted once solved?

**Comment by BhaaLseN** [comment]
Reviewed 2 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by delroth** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoCommon/DriverDetails.h, line 55 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-K8Uxfwx85U8S1nD0qWM:-K8VDg0ZviDII6mRdHVw:790412538) ([raw file](https://github.com/dolphin-emu/dolphin/blob/104286dd26ba3fd48ad4c05ebf560e13e861c4a0/Source/Core/VideoCommon/DriverDetails.h#L55)):*
OK. I'm fine with making them not enum classes then. It makes me very sad though :(

Please fix the bug with it causing diffs on Broadwell though.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by Sonicadvance1** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoCommon/DriverDetails.h, line 55 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-K8Uxfwx85U8S1nD0qWM:-K8VEXaUswJs45BGS3O7:-497468049) ([raw file](https://github.com/dolphin-emu/dolphin/blob/104286dd26ba3fd48ad4c05ebf560e13e861c4a0/Source/Core/VideoCommon/DriverDetails.h#L55)):*
Fixed

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by BhaaLseN** [comment]
Reviewed 1 of 3 files at r2.
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoCommon/DriverDetails.h, line 55 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-K8Uxfwx85U8S1nD0qWM:-K8VEKRCNSawftjFh_yt:-1334689153) ([raw file](https://github.com/dolphin-emu/dolphin/blob/104286dd26ba3fd48ad4c05ebf560e13e861c4a0/Source/Core/VideoCommon/DriverDetails.h#L55)):*
^ I can live with that too.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by delroth** [comment]
Reviewed 1 of 3 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by BhaaLseN** [comment]
Reviewed 3 of 3 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by Sonicadvance1** [comment]
Reviewed 3 of 3 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518)*

**Comment by Sonicadvance1** [comment]
A lot of the times we track the bug independently and I keep a list of bugs on a google drive document.
Could definitely be handling the tracking better
https://docs.google.com/spreadsheets/d/1ZbuYefVawKlzqiCtu7IU1XiXmgqO3z826-wiD_nw6lY/edit#gid=0

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3518#-:-K8VJXORIx3QvyM6x64a:-1586949596)*

**Comment by sheepdestroyer** [comment]
Yep, your spreadsheet does list the bug's reference at freedesktop. However it doesn't link to this github entry. So there is no cross referencing going on either way.
I could possibly imagine a bug getting resolved in the driver but the corresponding fix in dolphin not being reverted.

**Comment by sheepdestroyer** [comment]
As reported at the beginning of this PR, Radeon are also blacklisted for a similar bug.
The fix for Radeon seems to be in mesa now : https://bugs.freedesktop.org/show_bug.cgi?id=93648

Unfortunately I do not seem to be able to find a similar PR/bug for the radeon blacklisting to report it there.

**Comment by sheepdestroyer** [comment]
The original dolphin bug entry is also not listed in your spreadsheet or in this PR
https://bugs.dolphin-emu.org/issues/9166

**Comment by sheepdestroyer** [comment]
The patch fixing the original bug has been committed to mesa master : http://cgit.freedesktop.org/mesa/mesa/commit/?id=9f2e22bf343b21d6b44e6a502f00a86d169f5ade
It has also been CCed to 11.0 and 11.1 stable trees
The next major version (should be called 11.2) and stable releases (that would be 11.1.2 and 11.0.10) should include it.

**Comment by sheepdestroyer** [comment]
On IntelsSNB HD 3000, built Mesa-master with included fix as well as dolphin-master with blacklist removed.
I then tried Metroid Prime and can confirm the visual bugs are not present even when looking at minimap

**Comment by sheepdestroyer** [comment]
Mesa 11.2 has been branched 4 days ago and includes the aforementioned fix : https://cgit.freedesktop.org/mesa/mesa/log/?h=11.2

**Pull request merged**